Add pullspec overrides data for related images in each bundle image

qixiang opened the pull request:
For each bundle image we're going to rebuild when operator/operand
images are shipped, pullspec overrides data is required, the general
workflow is: 

1. Get the current image NVRs in advisory, these are rebuilt images
2. Get the original image NVRs for rebuilt images
3. Get digests of original images
4. Get digests of rebuilt images
5. With digests of original images, we can search in all latest bundle
   images to get the bundles which have original images are connected.
6. For each bundle image, if this image is not enabled for rebuild, skip
   it.
7. For each bundle image, if a related image is in original digests (step
   3), and osbs-pinning is used, update the digest with digest of rebuilt
   image.

JIRA: CLOUDWF-2902

apaplaus reviewed (commented):
LGFM except for some little comments, but I would still wait for Matt and for what he is famous for.... _his review_

qixiang commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -96,15 +127,29 @@ def handle(self, event):
+
+            for bundle in bundles:
+                bundle_digest = bundle.get('bundle_path_digest')
+                if not bundle_digest:
```
ok, I checked bundles data in pyxis, all bundles in redhat registry have `bundle_path_digest`, quay.io doesn't have such field, so yes, I agree we can remove this, for quay.io bundles, that's another story if we need to care about.

qixiang commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -115,6 +160,27 @@ def handle(self, event):
+                rebuilt_digest = rebuilt_digests_by_nvr[nvrs_mapping[image_nvr]]
+
+                pullspec_overrides_by_nvr.setdefault(bundle_nvr, {})
+                pullspec_overrides_by_nvr[bundle_nvr][original_digest] = rebuilt_digest
```
This is because we're adding multiple `original_digest: rebuilt_digest` to a single bundle image within the loop of 
```
for image_nvr, digests in bundle_digests_by_related_nvr.items():
```
multiple rebuilt images in an advisory can connect to a same bundle image, this is a common case.

apaplaus replied:
Oh ok, now I got it. Thx

qixiang commented:
Hi @mprahl could you help to review this?

mprahl commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -70,23 +70,54 @@ def handle(self, event):
+                )
+
+        if not original_digests_by_nvr:
+            msg = f"None of the original images have digest in Pyxis: {','.join(original_nvrs)}"
```
```suggestion
            msg = f"None of the original images have digests in Pyxis: {','.join(original_nvrs)}"
```

mprahl commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -70,23 +70,54 @@ def handle(self, event):
+                )
+
+        if not rebuilt_digests_by_nvr:
+            msg = f"None of the rebuilt images have digest in Pyxis: {','.join(rebuilt_nvrs)}"
```
```suggestion
            msg = f"None of the rebuilt images have digests in Pyxis: {','.join(rebuilt_nvrs)}"
```

mprahl commented on freshmaker/pyxis.py:
```diff
@@ -190,48 +190,39 @@ def get_latest_bundles(self, index_images):
-                if image.get('digest') in original_digests:
-                    ret_bundles.append(bundle)
-                    break
+            related_digests = [img.get('digest') for img in bundle.get('related_images', [])]
+            if image_digest in related_digests:
```
This would avoid looping through all the related images when there is a match.
```suggestion
            if any (image_digest == img.get('digest') for img in bundle.get('related_images', [])):
```

mprahl commented on freshmaker/pyxis.py:
```diff
@@ -190,48 +190,39 @@ def get_latest_bundles(self, index_images):
         """
-        Filter out bundles that don't have at least one 'related_image'
-        with the same manifest list digest as those in 'original_digests'
+        Get latest bundles that have the specified image digest in related images.
```
I suggest this change in wording since the method does not determine what is latest, the input to is what determines that.
```suggestion
        Get bundles that have the specified image digest in related images.
```

mprahl commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -96,15 +127,23 @@ def handle(self, event):
-            bundle_digests.add(bundle['bundle_path_digest'])
+
+            for bundle in bundles:
+                bundle_digest = bundle.get('bundle_path_digest')
```
If I recall correctly, there is a bug that causes `bundle_path_digest` not to be set in the data returned by Pyxis. So we should guard against this like the previous code. If you've confirmed that this is no longer the case, then access the key directly to indicate that it is always set (i.e. don't use `.get`).

mprahl commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -115,6 +154,27 @@ def handle(self, event):
         bundles_nvrs = self._filter_bundles_by_pinned_related_images(
             auto_rebuild_nvrs)
 
+        # Now we have the bundle NVRs needs to be rebuilt, for each bundle NVR,
```
```suggestion
        # Now we have the bundle NVRs that need to be rebuilt, for each bundle NVR,
```

mprahl commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -96,15 +127,23 @@ def handle(self, event):
+                bundle_digest = bundle.get('bundle_path_digest')
+                bundle_digests_by_related_nvr.setdefault(image_nvr, []).append(bundle_digest)
+
+        bundle_digests = set()
```
It'd probably be safer to just skip the event if `bundle_digests_by_related_nvr` is empty at this point. We have no guarantee that the original NVRs of the images in the BOTAS advisory were ever part of a bundle image.

mprahl commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -115,6 +154,27 @@ def handle(self, event):
+        # construct the data of related images pullspec overrides
+        pullspec_overrides_by_nvr = {}
+        for bundle_nvr in bundles_nvrs:
+            images = [i for i in bundle_images if i['brew']['build'] == bundle_nvr]
```
Rather than loop through `bundle_images` for each `nvr`, could you loop through it once outside of this for loop and create a dictionary of NVR to image. Then this for loop can be changed to iterate through that dictionary.

mprahl commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -115,6 +154,27 @@ def handle(self, event):
 
+        # Now we have the bundle NVRs needs to be rebuilt, for each bundle NVR,
+        # construct the data of related images pullspec overrides
+        pullspec_overrides_by_nvr = {}
```
We need to end up with the same format that OSBS uses in the Brew build extra field of `image.operator_manifests.related_images.pullspecs`. You can see an example of this with the Brew build with the ID of 1464389.

It also needs to include the pullspecs that Freshmaker did not replace. So in summary, use the dictionary of `image.operator_manifests.related_images.pullspecs` as a starting point. Then update the references of the `new` fields for pullspecs that were updated to be the rebuilt digest instead of the vulnerable one.

mprahl commented:
@qixiang I think this PR could be simplified in general by starting from the original code and add a mapping of original_nvr digest to rebuilt_nvr digest (like you did) and then refactor `_filter_bundles_by_pinned_related_images` so that it returns a dictionary that has a mapping of bundle NVR to `image.operator_manifests.related_images.pullspecs` (from the Brew build extra data).

From there, loop through all the items in the new dictionary just mentioned and loop through each `pullspecs` list (i.e. the value of the dictionary item). If one of the `new` fields contains a pullspec with the digest of an `original_nvr`, replace just the digest with the new digest of the `rebuilt_nvr`.

The dictionary now contain the pullspec replacement metadata that OSBS needs.

mprahl commented on freshmaker/migrations/versions/fcba8824bf8d_add_bundle_pullspec_overrides.py:
```diff
@@ -0,0 +1,26 @@
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Could you remove these automatic comments please?

qixiang commented:
Hi @mprahl @apaplaus I refactored the code in this handler and pyxis module, and updated the patch, please help to review this.

The APIs are refactored because they return a overall list of images or NVRs without the mapping of digest to images, now we get the bundles which the images in advisory connect to first, then for each bundle, check auto rebuild tag and osbs pinning, finally for each bundle we're going to rebuild, update the digest in pullspec (`new`) to digest of rebuilt images. 

And other changes:
1. `get_images_by_digests` -> `get_images_by_digest`: 
- Instead of getting a overall list of images, we need to know the mapping of digest to images, so refactor it to call with a single digest.
- `manifest_list_digest` is used in bundle's related images, but for bundles, `manifest_schema2_digest` is used, so we need to get image by `manifest_list_digest` **or** `manifest_schema2_digest`.

2. Get brew extra metadata: `buildinfo.get("build", {}).get("extra", {})` -> `buildinfo.get("extra", {})`, there is no `build` key.
3. Get repository auto rebuild tags via endpoint `repositories/registry/{registry}/repository/{repository}` instead of `repositories`.

apaplaus reviewed (commented):
Except few small comments it seems fine

qixiang commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -70,23 +71,59 @@ def handle(self, event):
+                rebuilt_digests_by_nvr[nvr] = digest
+            else:
+                log.warning(
+                    f"Image manifest_list_digest not found for {nvr} in Pyxis, skip this image"
```
Updated.

qixiang commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -96,63 +133,144 @@ def handle(self, event):
+            )
+            bundle_data['osbs_pinning'] = related_images.get('created_by_osbs', False)
+            # Save the original pullspecs
+            bundle_data['pullspecs'] = related_images.get('pullspecs', [])
```
I don't think it can happen, every published bundle image should has pullspecs of related images, and even with that can happen, we'll skip this bundle in later for loop:
```
    for pullspec in bundle_data['pullspecs']:
    ...
```

mprahl commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -70,23 +71,61 @@ def handle(self, event):
+            return []
+
+        # Convert to mapping of digest to NVR for later use
+        original_nvrs_by_digest = {v: k for k, v in original_digests_by_nvr.items()}
```
Optional: You could just build this dictionary as part of the `for nvr in original_nvrs:` loop to avoid looping through the values again. It also makes it easier to read in my opinion.

mprahl commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -96,63 +135,144 @@ def handle(self, event):
+                image_digest, all_bundles
+            )
+            if not bundles:
+                log.info(f"No latest bundle image found for {image_nvr}")
```
```suggestion
                log.info(f"No latest bundle image with the related image of {image_nvr}")
```

mprahl commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -96,63 +135,144 @@ def handle(self, event):
-                log.warning("Bundle %s doesn't have 'bundle_path_digests' set",
-                            bundle['bundle_path'])
+        for digests in bundle_digests_by_related_nvr.values():
+            bundle_digests.update(digests)
```
Optional: You could populate this set at the same time that you populate `bundle_digests_by_related_nvr`. It avoid an additional loop over this dictionary and it makes it easier to read.

mprahl commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -96,63 +135,144 @@ def handle(self, event):
+            'pullspecs': [],
+        }
+
+        # Get images for each bundle digest, a bundle digest can has multiple images
```
```suggestion
        # Get images for each bundle digest, a bundle digest can have multiple images
```

mprahl commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -96,63 +135,144 @@ def handle(self, event):
+
+                if old_digest not in original_nvrs_by_digest:
+                    # This related image is not one of the original images
+                    continue
```
This check should happen before `pullspec['original'] = old_pullspec` is called since we want the original pullspec replacement that OSBS computed to remain for related images which weren't rebuilt.

mprahl commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -96,63 +135,144 @@ def handle(self, event):
+                # A pullspec path is in format of "registry/repository@digest"
+                old_pullspec = pullspec.get('new')
+                # Replace 'original' with the old 'new' pullspec
+                pullspec['original'] = old_pullspec
```
We don't want to modify this value since the original value remains in the bundle image metadata (e.g. the pullspec with the floating tag) and we want OSBS to be able to identify that original value and replace it with either the rebuilt image or what it was previously pinned to.

mprahl commented:
@qixiang this is looking really good. Just a few comments to address. Afterwards I'll review the code one more time and review the tests.

qixiang commented:
Thank you @mprahl , all comments have been addressed in updated PR.
